NXO: Docs (#74)
paypal/iOS-SDK: docs/PayPalNativeCheckout/README.md

# Accepting PayPal Native Checkout Payments

The PayPal Native Checkout module in the PayPal SDK enables PayPal payments via a native UI experience in your app.

Follow these steps to add PayPal Native Checkout payments:

1. [Setup a PayPal Developer Account](#setup-a-paypal-developer-account)
1. [Add PayPal Native Checkout Module](#add-paypal-native-checkout-module)
1. [Test and go live](#test-and-go-live)

## Setup a PayPal Developer Account

You will need to set up authorization to use the PayPal Payments SDK. 
Follow the steps in [Get Started](https://developer.paypal.com/api/rest/#link-getstarted) to create a client ID. 

You will need a server integration to create an order to capture funds using the [PayPal Orders v2 API](https://developer.paypal.com/docs/api/orders/v2). 
For initial setup, the `curl` commands below can be used as a reference for making server-side RESTful API calls.

> 

Reviewer: We should also mention in this step that they need to fetch an ACCESS_TOKEN from their server. All good if we want this in a separate PR, since it affects all docs.

Author: @scannillo this is mentioned in the [general readme](https://github.com/paypal/iOS-SDK/blob/main/README.md) before we redirect them to each module.

Author: maybe we should add it to each module? but we would be repeating ourselves

## Add PayPal Native Checkout Module

### 1. Add the PayPal Native checkout to your app

#### Swift Package Manager

In Xcode, add the PayPal SDK as a [package dependency](https://developer.apple.com/documentation/swift_packages/adding_package_dependencies_to_your_app) to your Xcode project. Enter https://github.com/paypal/iOS-SDK as the package URL. Tick the "PayPalNativeCheckout" checkbox to add the PayPal Native Checkout package to your app.

In your app's source code files, use the following import syntax to include the PayPal Native Checkout module:

```swift
import PayPalNativeCheckout
```

#### CocoaPods

Include the PayPal Native Checkout pod in your `Podfile`.

```ruby
pod 'PayPalNativeCheckout'
```

In your app's source files, use the following import syntax to include PayPal's libraries:

```swift
import PayPalNativeCheckout
```

### 2. Initiate the Payments SDK

Create a `CoreConfig` using an [access token](../../README.md#access-token):

```swift
let config = CoreConfig(accessToken: "<ACCESS_TOKEN>", environment: .sandbox)
```

> Create a `PayPalClient` to approve an order with a PayPal payment method:

Reviewer: I thought [we decided](https://braintree.slack.com/archives/C03U9SEQYHK/p1663773329402289?thread_ts=1663772091.281979&cid=C03U9SEQYHK) to rename this to `PayPalNativeCheckoutClient` to align with web better? Do we have JIRAs to do that work?

Author: Yep i missed the thread on slack, going to change it in this PR.

# Accepting PayPal Native Checkout Payments

> The PayPal Web Checkout module in the PayPal SDK enables PayPal payments via a webview in your app.

Reviewer 2: ```suggestion
The PayPal Native Checkout module in the PayPal SDK enables PayPal payments via a webview in your app.
```

> The PayPal Native Checkout module in the PayPal SDK enables PayPal payments via a webview in your app.

Reviewer: ```suggestion
The PayPal Native Checkout module in the PayPal SDK enables PayPal payments via a native UI experience in your app.
```

> Follow these steps to add PayPal Web Checkout payments:

Reviewer: ```suggestion
Follow these steps to add PayPal Native Checkout payments:
```

1. [Setup a PayPal Developer Account](#setup-a-paypal-developer-account)
1. [Add PayPal Native Checkout Module](#add-paypal-native-checkout-module)
1. [Test and go live](#test-and-go-live)

## Setup a PayPal Developer Account

You will need to set up authorization to use the PayPal Payments SDK. 
Follow the steps in [Get Started](https://developer.paypal.com/api/rest/#link-getstarted) to create a client ID. 

You will need a server integration to create an order to capture funds using the [PayPal Orders v2 API](https://developer.paypal.com/docs/api/orders/v2). 
For initial setup, the `curl` commands below can be used as a reference for making server-side RESTful API calls.

## Add PayPal Native Checkout Module

### 1. Add the PayPal Native checkout to your app

#### Swift Package Manager

In Xcode, add the PayPal SDK as a [package dependency](https://developer.apple.com/documentation/swift_packages/adding_package_dependencies_to_your_app) to your Xcode project. Enter https://github.com/paypal/iOS-SDK as the package URL. Tick the "PayPalNativeCheckout" checkbox to add the PayPal Native Checkout package to your app.

> In your app's source code files, use the following import syntax to include the PayPal Native Checkout module:
> 
> ```swift
> import PayPalNativeCheckout
> ```
> 

Reviewer: ```suggestion
```

Same _take it or leave it_ comment as below

Author: Same as above!

#### CocoaPods

Include the PayPal Native Checkout pod in your `Podfile`.

```ruby
pod 'PayPalNativeCheckout'
```

> In your app's source files, use the following import syntax to include PayPal's libraries:
> 
> ```swift
> import PayPalNativeCheckout
> ```

Reviewer: ```suggestion
```

_Take it or leave it comment -_ I think we can omit the actual `import` statement instructions from the docs. Including instructions on how to include the module via SPM and CocoaPods should suffice for an iOS dev ([example](https://developer.paypal.com/braintree/docs/guides/apple-pay/client-side/ios/v5#custom-ui-integration) in Braintree developer docs).

Reviewer: Sounds good!

### 2. Initiate the Payments SDK

Create a `CoreConfig` using an [access token](../../README.md#access-token):

```swift
let config = CoreConfig(accessToken: "<ACCESS_TOKEN>", environment: .sandbox)
```

Create a `PayPalClient` to approve an order with a PayPal payment method:

```swift
let payPalClient = PayPalClient(config: config)
```

### 3. Create an order

> 

Reviewer: ```suggestion
```

When a user initiates a payment flow, call `v2/checkout/orders` to create an order and obtain an order ID:

**Request**
```bash
curl --location --request POST 'https://api.sandbox.paypal.com/v2/checkout/orders/' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <ACCESS_TOKEN>' \
--data-raw '{
    "intent": "<CAPTURE|AUTHORIZE>",
    "purchase_units": [
        {
            "amount": {
                "currency_code": "USD",
                "value": "5.00"
            }
        }
    ]
}'
```

**Response**
```json
{
   "id":"<ORDER_ID>",
   "status":"CREATED"
}
```

The `id` field of the response contains the order ID to pass to your client.

> 

Reviewer: ```suggestion
```

```swift
payPalClient.start { createOrderAction in
    createOrderAction.set(orderId: orderID)
}
```

Implement `PayPalDelegate` to listen for result notifications from the SDK. In this example, we implement it in a view model:

```swift
extension MyViewModel: PayPalDelegate {

    func paypal(_ payPalClient: PayPalClient, didFinishWithResult approvalResult: Approval) {
        // order was successfully approved and is ready to be captured/authorized (see step 5)
    }

    func paypal(_ payPalClient: PayPalClient, didFinishWithError error: CoreSDKError) {
        // handle the error by accessing `error.localizedDescription`
    }

    func paypalDidCancel(_ payPalClient: PayPalClient) {
        // the user cancelled
    }

    func paypalWillStart(_ payPalClient: PayPalClient) {
        // the paypal pay sheet is about to appear. Handle loading views, spinners, etc.
    }

    func paypalDidShippingAddressChange(
        _ payPalClient: PayPalClient,
        shippingChange: ShippingChange,
        shippingChangeAction: ShippingChangeAction
    ) {
        // called when the user decides to change the address or the shipping method of the order.
    }
}
```

> For a working example please refer to [PayPalViewModel](../../Demo/Demo/ViewModels/PayPalViewModel.swift) in out Demo application

Reviewer 2: ```suggestion
For a working example please refer to [PayPalViewModel](../../Demo/Demo/ViewModels/PayPalViewModel.swift) in our Demo application
```
